File: solutions/php/0977-square-of-a-sorted-array.php
<?php

class Solution
{
    /**
     * @param int[] $nums
     * @return int[]
     */
    public function sortedSquares(array $nums): array
    {
        $i = 0;
        $j = count($nums) - 1;
        $index = $j;
        $squares = array_fill(0, count($nums), 0);

        while ($index >= 0) {
            $iSqrt = $nums[$i] ** 2;
            $jSqrt = $nums[$j] ** 2;

            if ($iSqrt > $jSqrt) {
                $squares[$index] = $iSqrt;
                $i++;
            } else {
                $squares[$index] = $jSqrt;
                $j--;
            }
            $index--;
        }

        return $squares;
    }
}

File: solutions/go/0977-square-of-a-sorted-array.go
package main

func sortedSquares(nums []int) []int {
	squares := make([]int, len(nums))
	left, right := 0, len(nums)-1
	index := right

	for index >= 0 {
		lSqrt := nums[left] * nums[left]
		rSqrt := nums[right] * nums[right]

		if lSqrt > rSqrt {
			squares[index] = lSqrt
			left++
		} else {
			squares[index] = rSqrt
			right--
		}
		index--
	}

	return squares
}
